### PR TITLE
ClassificationTask predict step

### PR DESCRIPTION
**Pull request by isaaccorley**
Running `trainer.predict()` with the `ClassificationTask` fails because the default pytorch_lightning `predict_step` expects a batch tuple and not a dict. This PR overrides `predict_step` to work with our classification datasets.

E.g.

```python
preds = trainer.predict(model=task, dataloaders=datamodule.test_dataloader())

# Or if your datamodule has a `predict_dataloader` method defined
preds = trainer.predict(model=task, datamodule=datamodule)
```

### CODE REVIEW

**Comment by calebrob6** [comment]
I don't understand the failing test here.

**Review by adamjstewart — COMMENTED** [review]
Should we add this to all trainers?

**Comment by adamjstewart** [comment]
No idea what's going on with the failing test, the latest passing commit on master has the same versions of mypy and numpy, so I don't know why it would be failing all of a sudden.

**Comment by adamjstewart** [comment]
Technically this could go in 0.3.2 since it doesn't actually add a new feature, it overrides an existing feature.

**Comment by isaaccorley** [comment]
Looks like it's a bug introduced with the latest mypy that breaks with Python 3.10.7. The fix will come out in the next mypy release. https://github.com/python/mypy/issues/13627

**Comment by isaaccorley** [comment]
The plan is to add this to all the trainers but each one will have it's own specific needs, e.g you may not want to accumulate predicted segmentation masks into a list in memory and would rather save them in some other way.

**Comment by adamjstewart** [comment]
That is unfortunate, I guess dependabot can't control exactly which Python version we test. We could pin to 3.10.6 if we want to get the tests working.

**Comment by isaaccorley** [comment]
@adamjstewart mypy v0.981 is out which includes a fix for this failing test. Should I update `requirements/tests.txt` in this PR or a separate one?

**Comment by adamjstewart** [comment]
Dependabot will update it for you if you wait < 24 hrs (I think it checks once per day).

**Comment by isaaccorley** [comment]
Not sure why codecov is complaining. When I go to the link it only shows coverage dips on scripts I didn't modify.

File: torchgeo/trainers/classification.py [hunk 1]
```diff
@@ -238,6 +238,20 @@ def test_epoch_end(self, outputs: Any) -> None:
...
+        Args:
+            batch: the output of your DataLoader
+
+        Returns:
+            predicted softmax probabilities
+        """
+        batch = args[0]
+        x = batch["image"]
+        y_hat: Tensor = self(x).softmax(dim=-1)
+        return y_hat
```
Comment (adamjstewart) [inline torchgeo/trainers/classification.py:253]: Why is the predict step so different from the train/val/test steps? Shouldn't it look something like:
```suggestion
        y_hat = self.forward(x)
        y_hat_hard = y_hat.argmax(dim=1)
        return y_hat_hard
```

Comment (calebrob6) [reply torchgeo/trainers/classification.py:253]: Because you will frequently want the per-class probabilities instead of just the top class

**Review by calebrob6 — APPROVED** [review]

File: torchgeo/trainers/classification.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/trainers/classification.py:253]: And that's not true for train/val/test?

Comment (isaaccorley) [reply torchgeo/trainers/classification.py:253]: The train/val/test steps aren't meant to return data to a user. The operations that happen are only meant to perform the forward pass and compute metrics (which is why the argmax op is needed). The predict step needs to return the raw softmax outputs because we don't want to assume what a user wants to do with the output.

Comment (adamjstewart) [reply torchgeo/trainers/classification.py:253]: Do we need `yhat: Tensor` or can we just use `yhat`?

Comment (isaaccorley) [reply torchgeo/trainers/classification.py:253]: I had to add that to avoid mypy complaining.

Comment (adamjstewart) [reply torchgeo/trainers/classification.py:253]: Why was it complaining? If you use `self.forward(x)` instead does it stop complaining? What if you break it into multiple lines like we do in other steps?

Comment (adamjstewart) [reply torchgeo/trainers/classification.py:253]: (type hint warnings are not something to be ignored, they're something to be fixed)

Comment (calebrob6) [reply torchgeo/trainers/classification.py:253]: - The recommended way to run inference is with `self(x)` not `self.forward(x)` as `self(x)` calls hooks while `self.forward(x)` silently doesn't.
- `y_hat` is a Tensor so I don't see the problem. Breaking something into multiple lines to get `mypy` to stop complaining is ridiculous overhead.

Comment (adamjstewart) [reply torchgeo/trainers/classification.py:253]: Should we change other methods to replace `self.forward(x)` with `self(x)`? Just want to make sure we're consistent.

Comment (isaaccorley) [reply torchgeo/trainers/classification.py:253]: Yes we should change. We shouldn't be calling forward anywhere.

Comment (adamjstewart) [reply torchgeo/trainers/classification.py:253]: P.S. I feel like the `yhat: Tensor` thing is due to missing type hints for softmax but I couldn't figure out exactly where I need to add those. PyTorch's source code is a maze.

Comment (isaaccorley) [reply torchgeo/trainers/classification.py:253]: I think I was able to fix it by calling it as a function e.g. torch.softmax(tensor) instead of calling it as a method e.g. tensor.softmax()
